Implement ALPN/HTTP2

BenjaminLudwigSAP opened the pull request:
HTTP2 profile will be activated when listener has parameter `alpn_protocols` set. If that parameter does not contain HTTP2, the HTTP2 profile will behave like the already existing "normal" HTTP profile.
This PR will be merged together with other PRs in the milestone.

To do:
- [x] Create HTTP2 profile on F5s with name given in https://github.com/sapcc/helm-charts/pull/3861
- [x] Merge https://github.com/sapcc/helm-charts/pull/3861

BenjaminLudwigSAP committed:
[as3_restclient] Reorder protocol check, add comments

BenjaminLudwigSAP committed:
[as3_restclient] Add profile_http2 option

This will work as soon as
https://github.com/sapcc/helm-charts/pull/3861 is merged

BenjaminLudwigSAP committed:
[as3_restclient] Use HTTP2 profile if ALPN protocols set on listener

m-kratochvil commented:
Depending on the outcome of the design/implementation phase, we may need one shared HTTP2 profile, or we may need/want to implement two separate profiles - one for ingress and one for egress. I'm not sure about the impact on this PR should we decide to go with the second option.

BenjaminLudwigSAP committed:
[as3_restclient] Disable TLS renegotation for HTTP2 listeners

BenjaminLudwigSAP commented:
We have agreed in a meeting to first go with a shared HTTP2 profile and only maybe change it later.

BenjaminLudwigSAP committed:
[as3_restclient] Listener is HTTP2 only if h2 in alpn_protocols

It would suffice to count a listener with alpn_protocols set to
anything as HTTP2(-capable), because the HTTP2 profile acts like
normal HTTP when HTTP2 isn't explicitely requested via ALPN. However,
having a listener be able to speak HTTP2 without it being declared
that way is unexpected behavior.

BenjaminLudwigSAP commented:
(The five commits above are @velp rebasing on master)

m-kratochvil commented:
I have deployed the shared http/2 profile `cc_http2_profile` to all lbaas Big-IPs in qa-de-1.

velp commented:
@m-kratochvil in your JSON examples from F5 I saw the profile `cc_http2_profile` shoulbe in `profileHTTP2` section:
```
"listener_237e96aa-8bb6-4284-9141-ae1613162911": {
    "virtualAddresses": [
        "10.180.88.99"
    ],
    "virtualPort": 443,
    "persistenceMethods": [],
    "iRules": [],
    "policyEndpoint": [],
    "label": "test-listener-3",
    "serverTLS": "tls_listener_237e96aa-8bb6-4284-9141-ae1613162911",
    "redirect80": false,
    "pool": "pool_50fb3f12-aae9-44dd-9840-594b96d8fdbf",
    "profileHTTP": {
        "bigip": "/Common/cc_http_profile"
    },
    "profileTCP": {
        "bigip": "/Common/cc_tcp_profile"
    },
    "profileHTTP2": {
        "bigip": "/Common/cc_http2_profile"
    },
    "snat": "self",
    "class": "Service_HTTPS"
},
```
but in our plugin we will use `profileHTTP` (see this line: https://github.com/sapcc/octavia-f5-provider-driver/pull/170/files#diff-ab11f6353c7fb4c612c47ccd7e59aae5da90959519478c0a1e4d99890f5c815dR337).

Will it work?

m-kratochvil commented:
Not sure if I correctly understand the question, but for http2 enabled listener, both `profileHTTP` and `profileHTTP2` must be used, they complement each other. Basically, the profiles must look exactly as in the example you posted.
```
    "profileHTTP": {
        "bigip": "/Common/cc_http_profile"
    },
    "profileTCP": {
        "bigip": "/Common/cc_tcp_profile"
    },
    "profileHTTP2": {
        "bigip": "/Common/cc_http2_profile"
    },
```
To be more specific, the `cc_http_profile` is the "main" http profile for a listener, it handles http headers, cookies etc., while the `cc_http2_profile` adds ALPN support and some performance tweaks.

velp commented:
I mean, in the current solution (in this MR) we will have JSON like:
```
    "profileHTTP": {
        "bigip": "/Common/cc_http2_profile"
    },
    "profileTCP": {
        "bigip": "/Common/cc_tcp_profile"
    },
```
should it be changed to:
```
    "profileHTTP": {
        "bigip": "/Common/cc_http_profile"
    },
    "profileTCP": {
        "bigip": "/Common/cc_tcp_profile"
    },
    "profileHTTP2": {
        "bigip": "/Common/cc_http2_profile"
    },
```
?

m-kratochvil commented:
Yes, it needs to be changed to look like that.

Configuration
```
"profileHTTP": {
        "bigip": "/Common/cc_http2_profile"
    }
```
is not compatible, and should fail the declaration parsing.

velp committed:
Use correct iRule profile for HTTP/2. Add error logging for misconfiguration.

velp committed:
Fix typo
